### ISSUE TITLE
Use send Async instead of send in eth_getProof RPC call

### ISSUE THREAD

**Issue by 0xsarvesh** [OpenST/mosaic.js#158]
https://github.com/OpenST/mosaic.js/blob/1eb44c65e78cdd20c71009d30721d41c2642fc9a/src/utils/ProofGenerator.js#L117

Some provider like web3-provider-engine doesn't support `send`. In order to support more clients, mosaic.js should use `sendAsync` instead of `send`.  

Current implementation
```
     web3.currentProvider.send(
        {
          jsonrpc: '2.0',
          method: 'eth_getProof',
          params,
          id: new Date().getTime(),
        },
```

Required implementation

```
     web3.currentProvider.sendAsync(
        {
          jsonrpc: '2.0',
          method: 'eth_getProof',
          params,
          id: new Date().getTime(),
        },
```

**Comment by abhayks1** [comment]
Before going for the implementation of this ticket I spent time analyzing if we should do above change or not.

**Context**

In [JLP](https://github.com/OpenST/JLP) we have intercepted transaction signing using ZeroClientProvider.  ZeroClientProvider uses [web3-provider-engine](https://www.npmjs.com/package/web3-provider-engine) npm as ProviderEngine. web3-provider-engine has no implementation of web3.currentProvider.send. While generating proof, **web3.currentProvider.send** throws error.

**Problem**

web3-provider-engine has implemented sendAsync but throws error when calling send.
Error: ``` Web3ProviderEngine does not support synchronous requests.```

Web3 however has removed sendAsync method both from HttpProvider as well as WebsocketProvider.
Refer: https://github.com/ethereum/web3.js/issues/1451

**Solution**

In src/connection/index.js where we create web3 connections for origin and auxiliary chain we can do below:
```
originWeb3.currentProvider.send = originWeb3.currentProvider.sendAsync;
auxiliaryWeb3.currentProvider.send = auxiliaryWeb3.currentProvider.sendAsync;
```
The solution basically defines implementation of send function which is same implementation of sendAsync function.

cc: @benjaminbollen @schemar @sarvesh-ost 

Let me know if there is better solution.
We can create a separate ticket for the solution implementation.

**Comment by schemar** [comment]
* The [web3-provider-engine](https://github.com/MetaMask/web3-provider-engine) provides `sendAsync`, but we should not assume MetaMask's web3-provider-engine to be the provider.
* I don't think your suggested change helps, as `send` and `sendAsync` should have different behavior (as they have different names).

In my opinion, it looks like it cannot be fixed. Does this mean we cannot get the proof from within [OpenST/JLP](https://github.com/OpenST/JLP)?

**Comment by schemar** [comment]
🧐 We could just use whatever is there. This is JavaScript and functions are "objects".
Something like this maybe (not tested) (hold back gag reflex):
```js
web3.currentProvider[web3.currentProvider.sendAsync ? 'sendAsync' : 'send'](
  {
    jsonrpc: '2.0',
    method: 'eth_getProof',
    params,
    id: new Date().getTime(),
  },
  ...
```
This should use `sendAsync` if it exists on the provider and `send` if it does not.

**Comment by schemar** [comment]
Or more verbose:
```js
const sendFunctionIdentifier = web3.currentProvider.sendAsync ? 'sendAsync' : 'send';
web3.currentProvider[sendFunctionIdentifier](
  {
    jsonrpc: '2.0',
    method: 'eth_getProof',
    params,
    id: new Date().getTime(),
  },
  ...
```

**Comment by abhayks1** [comment]
> In my opinion, it looks like it cannot be fixed. Does this mean we cannot get the proof from within [OpenST/JLP](https://github.com/OpenST/JLP) ?

That's correct. We cannot get the proof from within [OpenST/JLP] without a fix.

**Comment by abhayks1** [comment]
> Or more verbose:
> 
> ```js
> const sendFunctionIdentifier = web3.currentProvider.sendAsync ? 'sendAsync' : 'send';
> web3.currentProvider[sendFunctionIdentifier](
>   {
>     jsonrpc: '2.0',
>     method: 'eth_getProof',
>     params,
>     id: new Date().getTime(),
>   },
>   ...
> ```

Thank you for an alternative approach. 
However considering web3 has deprecated .sendAsync function,  It's quite possible **web3-provider-engine** and other providers will deprecate sendAsync function soon in later versions. **HttpProvider** and **WebSocketProvider** has already removed sendAsync implementation.

I believe we should keep mosaic.js layer clean and refrain from quick patchy solution at that layer.
Considering JLP is a test framework,  I suggest we can apply patch solution at JLP layer for now to unblock JLP [PR#91](https://github.com/OpenST/JLP/pull/91) and move ahead.

What do you guys think?

**Comment by 0xsarvesh** [comment]
`web3.currentProvider.send` is available in zero providers but it simply throws an error saying `unimplemented method`. 
I also feel we should keep mosaic.js as it is. Whatever fix is needed should be done in JLP.

**Comment by abhayks1** [comment]
@benjaminbollen @schemar What do you think on the proposed solutions ?

**Comment by benjaminbollen** [comment]
Maybe someone can help me clarify ~~one~~ two questions:
1. is this specifically a problem for `getProof` and not for other calls? 
2. and how does Stake/Mint/Facilitator then work right now in JLP?

**Comment by abhayks1** [comment]
> Maybe someone can help me clarify two questions:
> 
> 1. is this specifically a problem for `getProof` and not for other calls?

Yes that's correct. That's because proofGenerator in mosaic.js uses **eth_getProof jsonrpc** call.
https://github.com/OpenST/mosaic.js/blob/develop/src/utils/ProofGenerator.js#L117
```
web3.currentProvider.send(
        {
          jsonrpc: '2.0',
          method: 'eth_getProof',
          params,
          id: new Date().getTime(),
        },
```
> 2. and how does Stake/Mint/Facilitator then work right now in JLP?

JLP smoke tests for Stake/Mint/Facilitator uses "ws" provider engine for constructing web3 object. "ws" provider implements .send function. "ws" provider has no implementation of .sendAsync function.

For running JLP executables, zero_client_provider is used for creating connection object. It requires "web3-provider-engine" as provider engine which throws error on calling .send function.
https://github.com/OpenST/JLP/blob/master/src/connection/zero_client_provider.js#L32

@sarvesh-ost @schemar Please add if I am missing any point.

**Comment by abhayks1** [comment]
Created an issue in MetaMask/web3-provider-engine repository. Hopefully they reply soon and we get more clarity.
https://github.com/MetaMask/web3-provider-engine/issues/309

**Comment by abhayks1** [comment]
**Update**

I did small POC test of client side signing by adding an account object in web3 wallet with RPC endpoint. It worked fine. With WS provider faced some issue which my feeling is fixable.

I believe this ticket can be closed since there needs to be follow up ticket in JLP to use web3 wallet for signing 🤔? Let me know the suggestions.

**Comment by benjaminbollen** [comment]
closing because irrelevant
